pbs_light.go

```diff
@@ -915,7 +949,7 @@ func serve(cfg *config.Configuration) error {
 		},
 	}
 	theMetrics := pbsmetrics.NewMetrics(metricsRegistry, openrtb_ext.BidderList())
-	theExchange := exchange.NewExchange(theClient, pbc.NewClient(&cfg.CacheURL), cfg, theMetrics)
+	theExchange := exchange.NewExchange(theClient, pbc.NewClient(&cfg.CacheURL), cfg, theMetrics, CurrencyRates.Load().([]byte))
```

Comment:
This will only update the currency rates when the server is restarted? So we either need to schedule regular restarts or the currency conversions will fall out of date.

Reply:
this will update the currency rates every 24 hours. please see the currencyFetchTicker above.

Reply:
But when the exchange is created, it is fed `CurrencyRates.Load().([]byte)`. So a byte slice representing the current (initial) set of exchange rates is created. I don't see how the updated value of the exchange rates will make it in to the exchange object once the 24 hour ticker ticks.

The way I would do it is add a pointer to `currencyData` into the exchange object. (It will have to be exported as `CurrencyData`. Add a method `UpdateCurrencyData(curData *CurrencyData)` that will overwrite the pointer with a new value. Then create the exchange object before kicking off the ticker goroutine. The goroutine will fetch the current currency rates, create a new blank `CurrencyData` object, and then unmarshall the currency rates into it. Once the unmarshalling is done without errors, it can call the update method on the exchange to update the pointer. This way we unmarshall once per 24 hours, and outside the request processing stream, rather than multiple times during a request with bid conversions.

Updating the pointer is an atomic operation so should be thread safe. Either the pointer is pointing to the old data or the new data, it would never be pointing to a partially processed structure. It is possible that for a request, some currency conversions would use the old rate, and others would use the new rate, but that is also possible using the atomic module, as it would just enforce that you don't read the value in the middle of an update.

Reply:
Ok, David just updated me that pointer assignment may not be atomic in go, so we may have to use `atomic.StorePointer()`.

Reply:
Digging into the details, technically a pointer assignment should be atomic. But go does do type safe things, so a "pointer" might include more, such as a description of the object it is pointing to. Also garbage collection needs to know how many references to that memory location exist. So all in all, I would expect that a naive pointer assignment would work 99% of the time, just that last 1% could potentially get as bad as crashing the server or overwriting random memory locations.

One alternative to `atomic` that I have though of was filling a channel with pointers to `CurrencyData`. However you will need to keep the channel full of pointers, and then switch to the new pointer once the new data is available. SHould be doable, but I can't come up with a way that would be superior to using `atomic.StorePointer()`

In my digging, I think we will need to use `atomic.LoadPointer()` to read the stored data into a local pointer variable in `convertCurrencyWithCurrentRates()` before accessing the currency data, to insure we are reading from a consistent pointer to the currency data.